src/main/kotlin/org/rust/lang/core/psi/ext/RsPat.kt

```diff
@@ -0,0 +1,13 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.psi.ext
+
+import org.rust.lang.core.psi.*
+
+val RsPat.isRefutable: Boolean get() = when (this) {
+    is RsPatTupleStruct, is RsPatRange, is RsPatConst, is RsPatStruct -> true
```

Comment:
You need to recursively check this property:

```rust
if let (1, 2)   = ... // refutable
if let (x, y)   = ... // irrefutable

if let [1, 2]   = ... // refutable
if let [x, y]   = ... // irrefutable

if let S { 42 } = ... // refutable
if let S { x }  = ... // irrefutable

if let S(42)    = ... // refutable
if let S(x)     = ... // irrefutable

if let box 42   = ... // refutable
if let box x    = ... // irrefutable
```

Reply:
Done.

Reply:
And also `RsPatRef`:

```
if let &42 = ... // refutable
if let &x  = ... // irrefutable
```

p.s. It would be nice to create tests from the above examples.

Reply:
Done.